metric/registry/registry.go

```diff
@@ -25,52 +25,88 @@ import (
 
 // MeterProvider is a standard MeterProvider for wrapping `MeterImpl`
 type MeterProvider struct {
-	impl metric.MeterImpl
+	lock   sync.Mutex
```

Comment:
After your feedback I studied this code in greater detail and realized more simplifications:

1. the metric/registry package was entirely internal, so moved it to internal/metric/registry
2. registry.MeterProvider was only being used in Controller, so I've moved it there, but this also requires an explanation:
2a. the global package did not use the registry MeterProvider because it has specific synchronization requirements and because it's also required to keep track of the meter details to perform its function. Note the global package still uses the unique instrument registry function, so there are still two users of the code in internal/metric/registry
2b. the metrictest package doesn't use the registry, "just because" the tests don't include uniqueness checking for meters or instruments.

As for `sync.Map`, I don't think there's a great need for concurrency. This lock would prevent concurrent instrument registration from collection activities, but I don't expect that in practice. To use the `sync.Map` requires a bit more work because you have to construct an object for `LoadAndStore()`, and since this object is trying to establish at-most-once semantics, it would require another lock somewhere.

Moving the MeterProvider functionality into the Controller simplified this PR a bit further. Please take another look specifically at the new implementation in controller.go.

Reply:
It's fine without the sync.Map.  The reason I asked was that Maps have a built-in at most one with [Map.Load**Or**Store](https://pkg.go.dev/sync#Map.LoadOrStore).  I get there is overhead because of the translation via interfaces, maybe this kind of change would be better suited when we can use generics.

Reply:
I was imagining the contract of the registry is that it will not call e.g. NewSyncInstrument() on the underlying implementation more than once. The aggregation pipeline expects a unique `*MetricDescriptor` for its purposes, so to use `LoadOrStore()` here I imagine an intermediate object, with a pattern like:

```
  newTmp := &intermediate{}
  result, loaded := controller.meters.LoadOrStore(key, newTmp)
  newTmp.initOnce.Do(func() {. // where initOnce is a sync.Once
    result.inst = // ... initialize the underlying instrument once
  }
  // use result.inst
```

IOW like ea7bc599 :grin:

Reply:
I was too hasty. Another commit is following. Also, this is more difficult to test.